pkgs/http/lib/src/base_request.dart

```diff
@@ -89,14 +89,33 @@ abstract class BaseRequest {
   bool _finalized = false;
 
   static final _tokenRE = RegExp(r"^[\w!#%&'*+\-.^`|~]+$");
+
   static String _validateMethod(String method) {
     if (!_tokenRE.hasMatch(method)) {
       throw ArgumentError.value(method, 'method', 'Not a valid method');
     }
     return method;
   }
 
-  BaseRequest(String method, this.url)
+  /// On upload progress callback.
```

Comment:
Could we provide a (broadcast) stream instead, instead of putting the callback directly into the class.

That's the traditional way to provide events, and the `on` name would be traditional for such a stream.
```dart
/// Upload progress notification
///
/// Sends events when upload progress information is
/// available.
///
/// ....
Stream<Progress> get onUploadProgress;
```
with a 
```dart
abstract class Progress {
  int progress;
  int? get total;
}
```
(Which I'd make a record when we get records, and document as something users *must not* implement or extend.)

Reply:
If you agree, I can commit to implement this solution.

Reply:
Unfortunately I found that this is not a viable solution. What will I return in the getter you defined to BaseRequest? `StreamController.stream`. 

A StreamController is required and if the developer does not want to listen to this progress, it creates an unnecessary load. In addition, since I need to know if the user has decided to listen for progress and this controller must be nullable.

If I could access the private values of `BaseRequest` from within the client (e.g `IOClient`), maybe I could come up with suitable solutions. Otherwise, two more properties are required, such as `BaseRequest.onUploadProgress` and `BaseRequest.uploadProgressController?`. I can guess that you don't want this visible to the developer.

Reply:
You can always create the controller when requested, and store it in a private variable:
```dart
  StreamController<Progress>? _controller;
  Stream<Progress> get onProgress => (_controller ??= StreamController(onCancel: () {
    _controller = null;
  })).stream;

  void _addProgress(int bytes, int? total) => _controller?.add(Progress(bytes, total));
```

This code has very little overhead when it's not used (a single `null` field and a single `null` check of that field whenever there is progress).
You can check whether `_controller` is `null` anywhere you currently check whether `onProgressCallback` is set.
 
If you need to access the internals inside `IOClient`, then you should just make the `onProgress` getter abstract and implement it in all the concrete subclasses. (I admit that the lack of `protected` access is annoying.)

Or add a 
```dart
void addProgressToRequest(Progress progress, BaseRequest request) {
  request._addProgress(progress);
}
```
to `base_progress.dart`, use it from the other libraries, but don't export it as part of the public interface of the package.
(That's a traditional way to share privates between libraries in the same package.)

Reply:
AFAIK, we are waiting on a `package:http` major version bump to review changes like this.

Is that true @natebosch ?

Reply:
> AFAIK, we are waiting on a `package:http` major version bump to review changes like this.

I haven't evaluated how difficult it will be to roll this internally. I do think it would need to wait for a major version bump.

I'm still a little unsure how meaningful these numbers are, or why it's important that we provide this API in this package instead of having the caller watch the stream progress.

Reply:
You are indeed right.  In my implementation for Altogic, onUploadProgress was useful because I split the bytes into chunks even if bytes weren't streamed(e.g. multipartfile.fromBytes). 

If it is not a problem for you, a solution is also possible, such as automatically increasing the number of stream parts in certain situations.